include/dlaf/matrix/internal/tile_future_manager.h

```diff
@@ -0,0 +1,220 @@
+      std::rethrow_exception(current_exception_ptr);
+    }
+  });
+}
+
+template <class ReturnTileType, class TileType>
+hpx::future<ReturnTileType> getTileFuture(hpx::future<TileType>& tile_future) noexcept {
+  hpx::future<TileType> old_future = std::move(tile_future);
+  hpx::promise<TileType> p;
+  tile_future = p.get_future();
+  return setPromiseTileFuture<ReturnTileType>(old_future, p);
+}
+
+template <class T, Device device>
+class TileFutureManager {
+public:
+  using TileType = Tile<T, device>;
+  using ConstTileType = Tile<const T, device>;
+
+  TileFutureManager() {}
+
+  TileFutureManager(TileType&& tile) : tile_future_(hpx::make_ready_future(std::move(tile))) {}
+
+  hpx::shared_future<ConstTileType> getReadTileSharedFuture() noexcept {
+    if (!tile_shared_future_.valid()) {
+      tile_shared_future_ = std::move(getTileFuture<ConstTileType>(tile_future_));
+    }
+    return tile_shared_future_;
+  }
+
+  hpx::future<TileType> getRWTileFuture() noexcept {
+    tile_shared_future_ = {};
+    return getTileFuture<TileType>(tile_future_);
+  }
+
+  void clearSync() {
+    tile_shared_future_ = {};
+    tile_future_.get();
+  }
+
+protected:
+  TileFutureManager(hpx::future<TileType>&& tile_future,
+                    hpx::shared_future<ConstTileType>&& tile_shared_future)
+      : tile_future_(std::move(tile_future)), tile_shared_future_(std::move(tile_shared_future)) {}
+
+  // The future of the tile with no promise set.
+  hpx::future<TileType> tile_future_;
+
+  // If valid, a copy of the shared future of the tile,
+  // which has the promise set to the promise for tile_future_.
+  hpx::shared_future<ConstTileType> tile_shared_future_;
+};
+
+enum class TileStatus { None = 0, Read = 1, RW = 2 };
+
+template <class T, Device device>
+class ViewTileFutureManager : private TileFutureManager<T, device> {
+  using TileManagerType = TileFutureManager<T, device>;
+
+public:
+  using typename TileManagerType::TileType;
+  using typename TileManagerType::ConstTileType;
+
+  ViewTileFutureManager() : tile_status_(TileStatus::None) {}
+
+  ViewTileFutureManager(TileManagerType& tile_manager, bool /* force_RW */) {
+    setUpRW(tile_manager);
+  }
+
+  ViewTileFutureManager(ViewTileFutureManager& tile_manager, bool force_RW) {
+    // RW tiles are fine in any case.
+    if (tile_manager.status() == TileStatus::RW) {
+      setUpRW(tile_manager);
+      return;
+    }
+
+    // Non-RW tiles are fine only if force_RW is false.
+    assert(!force_RW);
+
+    if (tile_manager.status() == TileStatus::Read)
+      tile_shared_future_ = tile_manager.getReadTileSharedFuture();
+
+    tile_status_ = tile_manager.status();
+  }
+
+  ViewTileFutureManager(ViewTileFutureManager&& tile_manager) noexcept
+      : TileManagerType(std::move(tile_manager)), tile_status_(tile_manager.tile_status_),
+        tile_promise_(std::move(tile_manager.tile_promise_)),
+        tile_shared_promise_(std::move(tile_manager.tile_shared_promise_)) {
```

Comment:
Hmm, isn't this a bug? You have already moved `tile_manager`, using it afterwards seems like an undefined behaviour?

Reply:
The first statement moves the data members of the base class, and then the data members of the derived class are moved later.